File: leetcode47.cpp
class Solution {
public:
    void recursion(vector<int> nums, int begin,  vector<vector<int> > &res) {
        if (begin == nums.size() -1) {
            res.push_back(nums);
            return;
        }
        for (int i = begin; i < nums.size(); i++) {
            if (begin != i && nums[begin] == nums[i]) continue;
            swap(nums[begin], nums[i]);
            recursion(nums, begin+1, res);
        }
    }
    vector<vector<int> > permuteUnique(vector<int> &nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int> >res;
        recursion(nums, 0, res);
        return res;
    }
};

File: leetcode47.java
public class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        boolean[] used = new boolean[nums.length];
        Arrays.sort(nums);
        backtracking(res,new ArrayList<Integer>(), nums,used);
        return res;
    }
    
    private void backtracking(List<List<Integer>> res, List<Integer> cur, int[] nums, boolean[] used){
        // end condition
        if(cur.size() == nums.length){
            res.add(new ArrayList(cur));
            return;
        }
        
        for(int i = 0; i < nums.length; i++){
            if(used[i])
                continue;
            if(i > 0 && nums[i] == nums[i-1] && !used[i-1])
                continue;
            used[i] = true;
            cur.add(nums[i]);
            backtracking(res,cur,nums,used);
            used[i] = false;
            cur.remove(cur.size()-1);
        }
    }
}
